### DISCUSSION TITLE
Speed up Warning propagation

### DISCUSSION THREAD

**Discussion by GregoryTravis** [enso-org/enso#10174] [Ideas and feedback]
I recently made a change to `Vector.Builder` to _always_ propagate warnings, which resulted in a measurable slowdown in some cases. (I then added a flag to disable warning propagation, for times when it is not needed.)

Since `Vector.Builder` is used everywhere, I spent some time on benchmarking and optimization, but concluded that it might be more useful to think about speed up Warning propagation in general.

Agenda:
- Do we think the warning system needs to be sped up?
- Do we think it's possible to speed it up?
- What are some ideas to do so?

I have only one idea for this:

**Don’t attach warnings to values**
- gather warnings during the evaluation of a node
- optionally attach values to the warnings (instead of warnings to the values)
- propagate a node’s warnings to downstream nodes (but do not propagate a value’s warnings to every downstream value)

Please add other ideas here.

**Comment by JaroslavTulach** [comment]
- #8682

**Reply by JaroslavTulach** [reply]
Another related issue that @hubertp reminded us about is
- https://github.com/oracle/graal/pull/4916

The issue wasn't originally created to speed things up. I have no estimates, how much performance we could gain. The original goal was to _improve consistency_ when calling into _hosted Java_ code.

**Comment by radeusgd** [comment]
> * propagate a node’s warnings to downstream nodes (but do not propagate a value’s warnings to every downstream value)

I'm not sure if I understand the distinction here. How would the propagation to 'downstream nodes' work in any other way than by attaching to values?

**Reply by JaroslavTulach** [reply]
The current _warnings system_ in Enso reminds me of **value taint tracking** as described for example in [Towards efficient, multi-language dynamic taint analysis](https://dl.acm.org/doi/abs/10.1145/3357390.3361028) - I do understand the implication of **value taint tracking**, I have a clear _mental model_ of it and there are scientific papers describing how it should work and how it can be implemented using Truffle. Except the [above](https://dl.acm.org/doi/abs/10.1145/3357390.3361028) also [nodeperf work](https://dl.acm.org/doi/10.1145/3178372.3179527) may be interesting.

> 
>    gather warnings during the evaluation of a node
>   optionally attach values to the warnings (instead of warnings to the values)
>    propagate a node’s warnings to downstream nodes
>  (but do not propagate a value’s warnings to every downstream value)

I have to admit, I _don't have a mental model_ of how this shall work and why. Some `ThreadLocal`?

**Comment by GregoryTravis** [comment]
Meeting summary:
- The first thing to do is #8682, since it shouldn't be too hard and may speed things up enormously
- Additional things to consider
    - Implement the wrapping in GraalVM itself (https://github.com/oracle/graal/pull/4916)
    - Doing propagation at a node granularity rather than at all function applications, possibly following [taint analysis](https://github.com/orgs/enso-org/discussions/10174#discussioncomment-9676527)

**Comment by JaroslavTulach** [comment]
Please note that 
- #10555

by @Akirathan has just been integrated. If the speedup isn't enough, then we need another brainstorming meeting.
